src/Shape.cpp

```diff
@@ -23,6 +24,80 @@ bool Shape::intersect(const Ray &, float t0, float t1, double& distance) const {
 }
 
 
+/** @brief Shape::contains() determines whether the shape intersects a sphere with center p.
+ *  @return bool True means the sphere intersects the shape.
+ *  @math http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.49.9172&rep=rep1&type=pdf
+ **/
+bool Shape::intersect(const Vector3& p, const double radius) const {
+    if (p.length()-radius > mesh_.getMaxRadius()){
+        return false;
+    }
+
+    if (radius > 0.0) {
+        const double radius2 = radius*radius;
+        // load triangles
+        const std::vector<geo::Vector3>& t_points = mesh_.getPoints();
+        const std::vector<TriangleI>& triangles_i = mesh_.getTriangleIs();
+        for(std::vector<TriangleI>::const_iterator it = triangles_i.begin(); it != triangles_i.end(); ++it) {
+            const Vector3 &v1 = t_points[it->i1_];
+            const Vector3 &v2 = t_points[it->i2_];
+            const Vector3 &v3 = t_points[it->i3_];
+
+            Vector3 e1 = v1 - v2;
+            Vector3 e2 = v2 - v3;
+            Vector3 e3 = v1 - v3;
+
+            // check endpoints
+            if ((v1-p).length2() < radius2) return true;
+            if ((v2-p).length2() < radius2) return true;
+            if ((v3-p).length2() < radius2) return true;
+
+            // check line segments
+            double d1 = (v1-p).length2();  // distance between v1 and p squared
+            double d2 = e1.dot(v1-p);  // dot product between e1 and v1-p
+            if (d2>0) {
```

Comment:
You do the same thing three times. Could you prevent this duplication?

Reply:
It is possible. I made it like this so that I can re use some measurements, saving an operation or two. But maybe it's better to have clarity over a marginal speed increase